lib/polish_number.rb

```diff
@@ -21,86 +21,181 @@ module PolishNumber
 
   MILLIONS = {:one => 'milion', :few => 'miliony', :many => 'milionów'}
 
+  BILLIONS = {:one => 'miliard', :few => 'miliardy', :many => 'miliardów'}
+
+  CENTS = [:auto, :no, :words, :digits]
+
   CURRENCIES = {
-    :PLN => {:one => 'złoty', :few => 'złote', :many => 'złotych'}
+    :NO => {:one => '', :few => '', :many => '',
+            :one_100 => 'setna', :few_100 => 'setne', :many_100 => 'setnych', :gender_100 => :she},
+    :PLN => {:one => 'złoty', :few => 'złote', :many => 'złotych',
+            :one_100 => 'grosz', :few_100 => 'grosze', :many_100 => 'groszy'},
+    :USD => { :one => 'dolar', :few => 'dolary', :many => 'dolarów',
+            :one_100 => 'cent', :few_100 => 'centy', :many_100 => 'centów'},
+    :EUR => { :one => 'euro', :few => 'euro', :many => 'euro', :gender => :it,
+            :one_100 => 'cent', :few_100 => 'centy', :many_100 => 'centów'},
+    :GBP => { :one => 'funt', :few => 'funty', :many => 'funtów',
+            :one_100 => 'pens', :few_100 => 'pensy', :many_100 => 'pensów'},
+    :CHF => { :one => 'frank', :few => 'franki', :many => 'franków',
+            :one_100 => 'centym', :few_100 => 'centymy', :many_100 => 'centymów'},
+    :SEK => { :one => 'korona', :few => 'korony', :many => 'koron', :gender => :she,
+            :one_100 => 'öre', :few_100 => 'öre', :many_100 => 'öre', :gender_100 => :it}
   }
 
-  def self.translate(number, options={})
+  def self.validate(number, options)
     if options[:currency] && !CURRENCIES.has_key?(options[:currency])
-      raise ArgumentError, "unknown :currency option '#{options[:currency].inspect}'. Choose one from: #{CURRENCIES.inspect}"
+      raise ArgumentError, "Unknown :currency option '#{options[:currency].inspect}'." +
+                  " Choose one from: #{CURRENCIES.inspect}"
     end
 
-    number = number.to_i
+    if options[:cents] && !CENTS.include?(options[:cents])
+      raise ArgumentError, "Unknown :cents option '#{options[:cents].inspect}'." +
+                  " Choose one from: #{CENTS.inspect}"
+    end
+
+    unless (0..999999999999).include? number
+      raise ArgumentError, 'number should be in 0..999999999999 range'
+    end
+    options
+  end
+
+  def self.translate(number, options={})
+
+    options = validate(number, options)
+
+    options[:cents] ||= :auto
+    number = number.to_i if options[:cents]==:no
+    formatted_number = sprintf('%015.2f', number)
+    currency = CURRENCIES[options[:currency] || :NO]
 
-    unless (0..999999999).include? number
-      raise ArgumentError, 'number should be in 0..999999999 range'
+    digits = formatted_number.chars.map { |char| char.to_i }
+    result = process_1_999999999999(digits[0..11], options, number, currency)
+
+    process_99_0(result, digits, options, formatted_number[-2..-1], currency)
+
+  end
+
+  def self.add_currency(name, hash)
+    CURRENCIES[name]=hash
```

Comment:
This isn't thread-safe. Perhaps we should use https://github.com/ruby-concurrency/thread_safe or immutable data structure?